### PR TITLE
Lomb-Scargle False Alarm Probabilities

### PR DESCRIPTION
**Pull request by jakevdp**
This PR adds computation of false alarm probabilities to the ``astropy.stats.LombScargle`` class. It's mainly derived from the code in [Baluev 2007](https://arxiv.org/abs/0711.0330), and includes four different methods: the Baluev approximation, the Davies upper-bound, the approximate independent frequencies method, and a bootstrap-based approach.

(Thanks to @astrohackweek for some time to focus on this!)

### CODE REVIEW

**Comment by jakevdp** [comment]
Fixed a couple issues in doctests... turns out they weren't being run locally (tried for a while with @eteq to debug, but couldn't figure out what was going on).

It looks like some other failures might be due to a 32/64 bit issue – do some of the travis tests run on 32 bit architectures?

**Comment by bsipocz** [comment]
@jakevdp - circleCI runs on 32bit, all the travis ones are 64bit!

**Comment by bsipocz** [comment]
I'm not exactly sure why some of the jobs run into errors at collection time, I'll look at the details tomorrow.

**Comment by jakevdp** [comment]
Another thing that's confusing me: why tests are failing on some 2.7 runs, but not others.

**Comment by jakevdp** [comment]
I need to make scipy dependency optional

File: astropy/stats/lombscargle/core.py [hunk 1]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
                              fit_mean=self.fit_mean,
                              nterms=self.nterms)
         return y_fit * get_unit(self.y)
+
+    def distribution(self, power, cumulative=False):
+        """Expected periodogram distribution under the null hypothesis
```
Comment (eteq) [inline astropy/stats/lombscargle/core.py]: missing period

File: astropy/stats/lombscargle/core.py [hunk 2]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        dH = 1 if self.fit_mean or self.center_data else 0
+        dK = dH + 2 * self.nterms
+        dist = statistics.cdf_single if cumulative else statistics.pdf_single
+        return dist(power, len(self.t), self.normalization, dH=dH, dK=dK)
+
+    def false_alarm_probability(self, power, method='baluev',
+                                samples_per_peak=5, nyquist_factor=5,
+                                minimum_frequency=None, maximum_frequency=None,
+                                method_kwds=None):
+        """False alarm probability of periodogram maxima under the null hypothesis
```
Comment (eteq) [inline astropy/stats/lombscargle/core.py]: missing period

File: astropy/stats/lombscargle/core.py [hunk 3]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+                                                  t=self.t, y=self.y, dy=self.dy,
+                                                  normalization=self.normalization,
+                                                  method=method,
+                                                  method_kwds=method_kwds)
+
+    def false_alarm_level(self, false_alarm_probability, method='baluev',
+                          samples_per_peak=5, nyquist_factor=5,
+                          minimum_frequency=None, maximum_frequency=None,
+                          method_kwds=None):
+        """Level of maximum at a given false alarm probability
```
Comment (eteq) [inline astropy/stats/lombscargle/core.py]: missing period

File: astropy/stats/lombscargle/statistics.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
+"""
+Utilities for computing periodogram statistics.
```
Comment (eteq) [inline astropy/stats/lombscargle/statistics.py]: maybe point out this is meant to be an internal module?  Just so no later maintainer thinks "hey, I should import this into the public namespace"

**Comment by eteq** [comment]
(Note that the above review is an add-on to an in-person review between me and @jakevdp)

**Comment by jakevdp** [comment]
I'm very confused by the CircleCI error: it's failing saying scipy is not installed, on a doctest that is wrapped in ``doctest-requires:: scipy``.

Any ideas?

**Comment by jakevdp** [comment]
Debugging this is hindered by the fact that I can't seem to run doctests locally... my only way to check is to push and have Travis run them. @eteq and I tried for about half an hour to figure out why they're not running, but didn't make any progress.

**Comment by pllim** [comment]
> Any ideas?

Is there some weird tab/space or different OS endlines mixed up in the file?

Another idea is to replace the period with colon in "highly-aliased observing patterns." I did a search for that directive in the repo and noticed a general trend of that directive, when it works, is nested under something that had a colon. Just a wild guess though.

**Comment by pllim** [comment]
> can't seem to run doctests locally

Any traceback? Was Erik able to reproduce it on his machine? Might be caused by the same thing that failed Circle CI, or might be something in your Python environment. Hard to tell without more info.

**Comment by bsipocz** [comment]
@pllim - I'm having the same doctest issue for a long time now. Spent a couple of hours to try to fix it a few months/weeks ago, but didn't find a solution so abandoned wasting much more time on it as we didn't see it on travis. Now that others run into it more often I'll try to debug it properly. Example traceback:

```
========================================================================================= test session starts =========================================================================================
platform darwin -- Python 3.5.3, pytest-3.2.0, py-1.4.34, pluggy-0.4.0

Running tests with Astropy version 3.0.dev19866.
Running tests in lib.macosx-10.10-x86_64-3.5/astropy/stats docs/stats.

Date: 2017-09-04T09:26:29

Platform: Darwin-14.5.0-x86_64-i386-64bit

Executable: /Users/bsipocz/.virtualenvs/astropy-dev/bin/python

Full Python Version: 
3.5.3 (default, Apr 29 2017, 10:12:19) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.13.1
Scipy: 0.19.1
Matplotlib: 2.0.2
h5py: 2.7.0
Pandas: 0.20.3
Cython: 0.25.2
Using Astropy options: remote_data: none.

rootdir: /private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-l_37t2hg, inifile: setup.cfg
plugins: arraydiff-0.2.dev0, xdist-1.15.0, pep8-1.0.6, mpl-0.7, mock-1.6.0, cov-2.4.0, hypothesis-3.8.3
collected 1128 items 

astropy/stats/bayesian_blocks.py .
astropy/stats/biweight.py .....
astropy/stats/circstats.py ......
astropy/stats/funcs.py ...
astropy/stats/info_theory.py ....
astropy/stats/sigma_clipping.py ..
astropy/stats/spatial.py .
astropy/stats/lombscargle/core.py .
astropy/stats/lombscargle/implementations/utils.py .
astropy/stats/lombscargle/implementations/tests/test_mle.py ........................
astropy/stats/lombscargle/implementations/tests/test_utils.py .............................................
astropy/stats/lombscargle/tests/test_lombscargle.py ................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
astropy/stats/tests/test_bayesian_blocks.py .......
astropy/stats/tests/test_biweight.py ..................
astropy/stats/tests/test_circstats.py .........
astropy/stats/tests/test_funcs.py ...........x................................s
astropy/stats/tests/test_histogram.py .........
astropy/stats/tests/test_info_theory.py ....
astropy/stats/tests/test_jackknife.py ...
astropy/stats/tests/test_sigma_clipping.py ...........
astropy/stats/tests/test_spatial.py ............
../docs/stats/circ.rst F
../docs/stats/index.rst F
../docs/stats/lombscargle.rst F
../docs/stats/ripley.rst F
../docs/stats/robust.rst F

============================================================================================== FAILURES ===============================================================================================
____________________________________________________________________________________ [doctest] docs/stats/circ.rst ____________________________________________________________________________________
ImportError while importing test module '/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-l_37t2hg/docs/stats/circ.rst'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ImportError: No module named 'circ'
___________________________________________________________________________________ [doctest] docs/stats/index.rst ____________________________________________________________________________________
ImportError while importing test module '/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-l_37t2hg/docs/stats/index.rst'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ImportError: No module named 'index'
________________________________________________________________________________ [doctest] docs/stats/lombscargle.rst _________________________________________________________________________________
ImportError while importing test module '/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-l_37t2hg/docs/stats/lombscargle.rst'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ImportError: No module named 'lombscargle'
___________________________________________________________________________________ [doctest] docs/stats/ripley.rst ___________________________________________________________________________________
ImportError while importing test module '/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-l_37t2hg/docs/stats/ripley.rst'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ImportError: No module named 'ripley'
___________________________________________________________________________________ [doctest] docs/stats/robust.rst ___________________________________________________________________________________
ImportError while importing test module '/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astropy-test-l_37t2hg/docs/stats/robust.rst'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ImportError: No module named 'robust'
========================================================================================== warnings summary ===========================================================================================
None
  [pytest] section in setup.cfg files is deprecated, use [tool:pytest] instead.

-- Docs: http://doc.pytest.org/en/latest/warnings.html
============================================================== 5 failed, 1121 passed, 1 skipped, 1 xfailed, 1 warnings in 20.63 seconds ===============================================================
```

File: docs/stats/lombscargle.rst [hunk 5]
```diff
@@ -415,6 +418,191 @@ For more information on the statistical properties of these normalizations,
...
+AstroPy provides four options for approximating the false alarm probability,
+which can be chosen using the ``method`` keyword:
+
+- ``method="baluev"`` (the default)
+  implements the approximation proposed by Baluev 2008 [8]_,
+  which employs extreme value statistics to compute an upper-bound of the false
+  alarm probability for the alias-free case. Experiments show that the bound is
+  also useful even for highly-aliased observing patterns.
+
+  .. doctest-requires:: scipy
```
Comment (larrybradley) [inline docs/stats/lombscargle.rst]: >I'm very confused by the CircleCI error: it's failing saying scipy is not installed, on a doctest that is wrapped in .. doctest-requires:: scipy.
>
> Any ideas?

@jakevdp  You need to remove the leading spaces before `.. doctest-requires:: scipy`.  This is why your doctests are failing.

File: docs/stats/lombscargle.rst [hunk 6]
```diff
@@ -415,6 +418,191 @@ For more information on the statistical properties of these normalizations,
...
+
+- ``method="bootstrap"`` implements a bootstrap simulation: effectively it
+  computes many Lomb-Scargle periodograms on simulated data at the same
+  observation times. The bootstrap approach can very accurately determine
+  the false alarm probability, but is very computationally expensive:
+  to estimate the level corresponding to a false alarm probability
+  :math:`P_{false}`, it requires on order :math:`n_{boot} \approx 10/P_{false}`
+  individual periodograms to be computed for the dataset.
+
+  .. doctest-requires:: scipy
```
Comment (larrybradley) [inline docs/stats/lombscargle.rst]: and here.  Note that this doctest is skipped below, so this `.. doctest-requires` is really unnecessary.

File: docs/stats/lombscargle.rst [hunk 7]
```diff
@@ -415,6 +418,191 @@ For more information on the statistical properties of these normalizations,
...
+
+    >>> ls.false_alarm_probability(power.max(), method='bootstrap')  # doctest: +SKIP
+    0.0030000000000000027
+
+- ``method="davies"`` is related to the Baluev method, but loses accuracy
+  at large false alarm probabilities
+
+  .. doctest-requires:: scipy
```
Comment (larrybradley) [inline docs/stats/lombscargle.rst]: and here.

File: docs/stats/lombscargle.rst [hunk 8]
```diff
@@ -415,6 +418,191 @@ For more information on the statistical properties of these normalizations,
...
+
+    >>> ls.false_alarm_probability(power.max(), method='davies')  # doctest: +FLOAT_CMP
+    0.0043311525763707216
+
+- ``method="naive"`` is a simplistic method based on the assumption that
+  well-separated areas in the periodogram are independent. In general, it
+  provides a very poor estimate of the false alarm probability and should
+  not be used in practice, but is included for completeness
+
+  .. doctest-requires:: scipy
```
Comment (larrybradley) [inline docs/stats/lombscargle.rst]: and here.

**Comment by larrybradley** [comment]
FWIW, I'm not having any problems running the doctests locally.

File: docs/stats/lombscargle.rst (hunk 5 above)
Comment (jakevdp) [reply docs/stats/lombscargle.rst]: But the doctest in question is part of a bulleted list... Won't that end the list if I remove these spaces?

**Comment by jakevdp** [comment]
> FWIW, I'm not having any problems running the doctests locally.

Yeah, @eteq couldn't reproduce my issue either, and there's no error message or any other clue about why they're not being run.

**Comment by jakevdp** [comment]
I'm thinking about just putting ``DOCTEST: +SKIP`` on all of these so that tests will pass.

File: docs/stats/lombscargle.rst (hunk 5 above)
Comment (pllim) [reply docs/stats/lombscargle.rst]: Perhaps not... There is similar usage in https://raw.githubusercontent.com/astropy/astropy/bdb4ec936fa789a24db28d0329e4a66ed10892ec/docs/convolution/kernels.rst

Comment (jakevdp) [reply docs/stats/lombscargle.rst]: OK – I'll give it a try!

**Comment by pllim** [comment]
But skipping doctest will defeat the purpose of writing doctest... 😅 

If you use conda, another possibility is to export your conda env, so someone could replicate the exact env on a similar hardware (if possible) and see... But like Brigitta said, as long as we can get CIs to pass, that should be good enough.

**Comment by jakevdp** [comment]
This code is tested elsewhere... the broken tests are in the narrative documentation, so skipping isn't as much of an issue. But I'm going to try de-denting before going the nuclear route :smile:

**Comment by bsipocz** [comment]
@jakevdp - Do you plan to work more on the docs, or it's good to go? While the approved button wasn't hit, I believe @larrybradley already did a review.

**Comment by jakevdp** [comment]
I'm pretty happy with this at this point. Let me know if you see any additional issues to be addressed

File: astropy/stats/lombscargle/core.py [hunk 9]
```diff
@@ -50,6 +51,9 @@ class LombScargle(object):
         of the input data. This is especially important if fit_mean = False
     nterms : int (optional, default=1)
         number of terms to use in the Fourier fit
+    normalization : string (optional, default='standard')
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: For a fixed list of possible values, the numpydoc style is to list them in braces with the default value first, i.e.:

`normalization : {'standard', 'model', 'log', 'psd'}, optional

File: astropy/stats/lombscargle/core.py [hunk 10]
```diff
@@ -50,6 +51,9 @@ class LombScargle(object):
...
+        Normalization to use for the periodogram.
+        Options are 'standard', 'model', 'log', or 'psd'.
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: And then this line is unnecessary.

File: astropy/stats/lombscargle/core.py [hunk 11]
```diff
@@ -191,6 +197,9 @@ def autofrequency(self, samples_per_peak=5, nyquist_factor=5,
         maximum_frequency : float (optional)
             If specified, then use this maximum frequency rather than one
             chosen based on the average nyquist frequency.
+        return_freq_limits : bool (optional)
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: We usually don't put "optional" in parens, but simply prepend with a comma (again based on the numpydoc style), e.g. :  `return_freq_limits : bool, optional`.  I do see several other instances of the `(optional)` usage in astropy, but most of them are in the stats modules that I think you wrote :smile:  (e.g. `bayesian_blocks.py` and `histogram.py`).  Leave it for now.

File: astropy/stats/lombscargle/core.py [hunk 12]
```diff
@@ -241,9 +253,8 @@ def autopower(self, method='auto', method_kwds=None,
 
         method_kwds : dict (optional)
             additional keywords to pass to the lomb-scargle method
-        normalization : string (optional, default='standard')
-            Normalization to use for the periodogram.
-            Options are 'standard', 'model', or 'psd'.
+        normalization : string (optional)
+            if specified, override the normalization specified at instantiation
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: "if" -> "If".  Also missing a period at the end.

File: astropy/stats/lombscargle/core.py [hunk 13]
```diff
@@ -303,9 +314,8 @@ def power(self, frequency, normalization='standard', method='auto',
             if True, assume that the input frequency is of the form
             freq = f0 + df * np.arange(N). Only referenced if method is 'auto'
             or 'fast'.
-        normalization : string (optional, default='standard')
-            Normalization to use for the periodogram.
-            Options are 'standard', 'model', 'log', or 'psd'.
+        normalization : string
+            if specified, override the normalization specified at instantiation
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: Same.

File: astropy/stats/lombscargle/core.py [hunk 14]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        This computes the expected probability distribution or cumulative
+        probability distribution of periodogram power, under the null
+        hypothesis of a non-varying signal with Gaussian noise. Note that
+        this is not the same as the expected distribution of peak values;
+        for that see the ``false_alarm_probability()`` method.
+
+        Parameters
+        ----------
+        power : array_like
+            the periodogram power at which to compute the distribution
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: same.

File: astropy/stats/lombscargle/core.py [hunk 15]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        cumulative : bool (optional)
+            if True, then return the cumulative distribution
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: same.

File: astropy/stats/lombscargle/core.py [hunk 16]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        """False alarm probability of periodogram maxima under the null hypothesis.
+
+        This gives an estimate of the false alarm probability given the height
+        of the largest peak in the periodogram, based on the null hypothesis
+        of non-varying data with Gaussian noise.
+
+        Parameters
+        ----------
+        power : array-like
+            the periodogram value
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: same

File: astropy/stats/lombscargle/core.py [hunk 17]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        maximum_frequency : float
+            the maximum frequency of the periodogram
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: same

File: astropy/stats/lombscargle/core.py [hunk 18]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        method : string
+            The approximation method to use; default='baluev'. Must be one of
+            ['baluev', 'davies', 'naive', 'bootstrap']
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: period at end

File: astropy/stats/lombscargle/core.py [hunk 19]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        method_kwds : dict (optional)
+            Additional method-specific keywords
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: period

File: astropy/stats/lombscargle/core.py [hunk 20]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+
+        Returns
+        -------
+        fap : np.ndarray
+            The false alarm probability
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: and period

File: astropy/stats/lombscargle/core.py [hunk 21]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        """Level of maximum at a given false alarm probability.
+
+        This gives an estimate of the periodogram level corresponding to a
+        specified false alarm probability for the largest peak, assuming a
+        null hypothesis of non-varying data with Gaussian noise.
+
+        Parameters
+        ----------
+        false_alarm_probability : array-like
+            the false alarm probability (0 < fap < 1)
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: capital and period

File: astropy/stats/lombscargle/core.py [hunk 22]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        maximum_frequency : float
+            the maximum frequency of the periodogram
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: capital and period

File: astropy/stats/lombscargle/core.py [hunk 23]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        method : string
+            The approximation method to use; default='baluev'. Must be one of
+            ['baluev', 'davies', 'naive', 'bootstrap']
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: period

File: astropy/stats/lombscargle/core.py [hunk 24]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+        method_kwds : dict (optional)
+            Additional method-specific keywords
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: period

File: astropy/stats/lombscargle/core.py [hunk 25]
```diff
@@ -356,3 +368,180 @@ def model(self, t, frequency):
...
+
+        Returns
+        -------
+        power : np.ndarray
+            The periodogram peak height corresponding to the specified
+            false alarm probability
```
Comment (larrybradley) [inline astropy/stats/lombscargle/core.py]: period

File: astropy/stats/lombscargle/tests/test_statistics.py [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+from ....tests.helper import pytest
+from .. import LombScargle
+from .._statistics import (cdf_single, pdf_single, fap_single, inv_fap_single,
+                           METHODS)
+from ..utils import convert_normalization, compute_chi2_ref
+
+METHOD_KWDS = dict(bootstrap={'n_bootstraps': 20, 'random_seed': 42})
+NORMALIZATIONS = ['standard', 'psd', 'log', 'model']
+
+def _require_scipy(method):
```
Comment (larrybradley) [inline astropy/stats/lombscargle/tests/test_statistics.py]: Is this function used anywhere?  If not, please remove it.

**Comment by jakevdp** [comment]
@larrybradley – in going over your suggested changes, I noticed a number of additional instances where capital letters and periods were missing in parameter descriptions. Is there a reason you asked me to correct some and not others? Is there some rule whereby you determine for parameter descriptions when capital letters and periods are required and when they are not?

**Comment by larrybradley** [comment]
@jakevdp  The standard is the parameter description should always begin with a capital letter, but I may have missed some in my review (or there may have been some already in the existing code, which I didn't review).  (sorry for the nits!)

**Comment by bsipocz** [comment]
Merging as this is now approved and passed all CI.

Thank you @jakevdp!

**Pull request merged**